libfwupdplugin/fu-pefile.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright (C) 2023 Richard Hughes <richard@hughsie.com>
+// SPDX-License-Identifier: LGPL-2.1+
+
+#[derive(Parse, Validate)]
+struct PeDosHeader {
+    magic: u16le: const=0x5A4D,
+    _cblp: u16le: default=0x90,
+    _cp: u16le: default=0x3,
+    _crlc: u16le: default=0x0,
+    _cparhdr: u16le: default=0x4,
+    _minalloc: u16le: default=0x0,
+    _maxalloc: u16le: default=0xFFFF,
+    _ss: u16le: default=0x0,
+    _sp: u16le: default=0xB8,
+    _csum: u16le: default=0x0,
+    _ip: u16le: default=0x0,
+    _cs: u16le: default=0x0,
+    _lfarlc: u16le: default=0x40,
+    _ovno: u16le: default=0x0,
+    _res: [u8; 8]: default=0x0000000000000000,
+    _oemid: u16le: default=0x0,
+    _oeminfo: u16le: default=0x0,
+    _res2: [u8; 20]: default=0x0000000000000000000000000000000000000000,
+    lfanew: u32le: default=0x80,
+}
+
+#[repr(u16le)]
+enum PeCoffMachine {
+    Unknown = 0x0,
+    Alpha = 0x184,
+    Alpha64 = 0x284,
+    Am33 = 0x1d3,
+    Amd64 = 0x8664,
+    Arm = 0x1c0,
+    Arm64 = 0xaa64,
+    Armnt = 0x1c4,
+    Axp64 = 0x284,
+    Ebc = 0xebc,
+    I386 = 0x14c,
+    Ia64 = 0x200,
+    Loongarch32 = 0x6232,
+    Loongarch64 = 0x6264,
+    M32r = 0x9041,
+    Mips16 = 0x266,
+    Mipsfpu = 0x366,
+    Mipsfpu16 = 0x466,
+    Powerpc = 0x1f0,
+    Powerpcfp = 0x1f1,
+    R4000 = 0x166,
+    Riscv32 = 0x5032,
+    Riscv64 = 0x5064,
+    Riscv128 = 0x5128,
+    Sh3 = 0x1a2,
+    Sh3dsp = 0x1a3,
+    Sh4 = 0x1a6,
+    Sh5 = 0x1a8,
+    Thumb = 0x1c2,
+    Wcemipsv2 = 0x169,
+}
+
+#[repr(u16le)]
+enum PeCoffMagic {
+    Pe32 = 0x10b,
+    Pe32Plus = 0x20b,
+}
+
+#[repr(u16le)]
+enum FuCoffSubsystem {
+    Unknown = 0,
+    Native = 1,
+    WindowsGui = 2,
+    WindowsCui = 3,
+    Os2Cui = 5,
+    PosixCui = 7,
+    NativeWindows = 8,
+    WindowsCeGui = 9,
+    EfiApplication = 10,
+    EfiBootServiceDriver = 11,
+    EfiRuntimeDriver = 12,
+    EfiRom = 13,
+    Xbox = 14,
+    WindowsBootApplication = 16,
+}
+
+#[derive(Parse)]
+struct PeCoffHeader64 {
+    signature: u32le: const=0x4550, // "PE\0\0"
+    machine: PeCoffMachine: default=0x8664,
+    number_of_sections: u16le: default=0,
+    _time_date_stamp: u32le: default=0,
+    pointer_to_symbol_table: u32le: default=0,
+    number_of_symbols: u32le: default=0,
+    size_of_optional_header: u16le: default=0xf0,
+    characteristics: u16le: default=0x2022,
+    magic: PeCoffMagic: default=0x020b,
```

Comment:
I would split the structure here to make it less confusing. That would also make `FU_STRUCT_PE_COFF_HEADER64_SIZE` the right size to add to `hdr_offset` (see my other comment).

Also in theory, some files do not have the optional header. So if one day we need that (though unlikely), it will be less confusing.

Reply:
Good idea, split. Is this version easier to understand?

Reply:
Perfect!